iotc-azrtos-sdk/src/iotconnect_di.c

```diff
@@ -0,0 +1,539 @@
+//
+// Copyright: Avnet, Softweb Inc. 2022
+// Created by Nik Markovic <nikola.markovic@avnet.com> on 5/15/22.
+//
+
+#include <stdlib.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <ctype.h>
+//TODO: remove remaining dependencies to TO
+#include "TO.h"
+#include "nx_crypto_sha2.h"
+#include "iotconnect_certs.h"
+#include "iotconnect_discovery_v3.h"
+#include "iotc_algorithms.h"
+#include "iotc_auth_driver.h"
+#include "azrtos_https_client.h"
+#include "iotconnect_device_identity.h"
+
+#define DDIM_AUTH_PATH "/x509/auth"
+#define DDIM_SIGN_PATH "/x509/cert/sign"
+#define DDIM_ACK_PATH "/x509/cert/ack"
+
+#define URL_RESOUCE_BUFFER_SIZE 40
+
+#define WORK_BUFFER_SIZE (TO_CERT_X509_MAXSIZE * 2 + 1)
+// Buffer that we will malloc and free when we are done.
+// Buffer will be used for conversions between binary and hex data.
+// 1 for null
+static uint8_t work_buffer[WORK_BUFFER_SIZE];
+static size_t work_buffer_size = 0;
+static char * discovery_agent_host = NULL;
+static char * discovery_agent_path = NULL;
+static char resource_buffer[URL_RESOUCE_BUFFER_SIZE];
+
+static char * bin_to_hex(uint8_t* bin_array, size_t bin_len) {
+	work_buffer_size = 0;
+	int hex_idx = 0;
+	for (int i = 0; i < bin_len; i++) {
+		work_buffer[hex_idx] = 0; // empty string
+		int ret = sprintf((char*)&work_buffer[hex_idx], "%02x", bin_array[i]);
+		if (ret != 2) {
+			printf("DDIM: bin_to_hex: buffer overflow!\r\n");
+			work_buffer[0] = 0;
+			return NULL;
+		}
+		hex_idx += 2;
+		if (hex_idx >= WORK_BUFFER_SIZE -1) {
+			printf("DDIM: bin_to_hex: buffer overflow!\r\n");
+			work_buffer[0] = 0;
+			return NULL;
+		}
+	}
+	work_buffer_size = hex_idx + 1; // for null;
+	return (char *) work_buffer;
+}
+static bool is_valid_hex_digit(char ch) {
+	return (ch >= '0' && ch <= '9') || (ch >= 'a' && ch <='f') || (ch >= 'A' && ch <= 'F');
+}
+int hex_digit_to_int(char digit) {
+    if (digit >= 'A' && digit <= 'F') {
+        return digit - 'A' + 10;
+    } else if (digit >= 'a' && digit <= 'f') {
+        return digit - 'a' + 10;
+    } else if (digit >= '0' && digit <= '9') {
+        return digit - '0';
+    }
+
+    return -1; // Bad input.
+}
+static uint8_t* hex_to_bin(char* hex_str) {
+	int bin_idx = 0;
+	int hex_idx = 0;
+	size_t hex_len = strlen(hex_str);
+
+	work_buffer_size = 0;
+
+	for (int i = 0; i < hex_len; i++) {
+		hex_str[i] = tolower(hex_str[i]);
+	}
+	while (hex_str[hex_idx] != 0) {
+		if (!is_valid_hex_digit(hex_str[hex_idx]) || !is_valid_hex_digit(hex_str[hex_idx+1])) {
+			printf("DDIM: hex_to_bin: Error: encountered non-hex characters or early end of string \"%c%c\"\r\n", hex_str[hex_idx], hex_str[hex_idx+1]);
+			return NULL;
+		}
+		int dmsb = hex_digit_to_int(hex_str[hex_idx]);
+		int dlsb = hex_digit_to_int(hex_str[hex_idx+1]);
+		if (dmsb == -1 || dlsb == -1) {
```

Comment:
if test#1 didn't fire will this ever happen?

Reply:
I'm not sure what tis is about. If test fails, we will return NULL indicating a non-valid result. If it does not, we will continue to assigning dmsb and dlsb. Both conditions seem to be valid to me.